File: 11_Graphs/12_Graph_Valid_Tree/0261-graph-valid-tree.cpp
/*
Problem: LeetCode 261 - Graph Valid Tree

Description:
You have a graph of n nodes labeled from 0 to n - 1. You are given an integer n and a list of edges where edges[i] = [ai, bi] indicates that there is an undirected edge between nodes ai and bi in the graph.
Return true if the edges of the given graph make up a valid tree, and false otherwise.

Intuition:
A valid tree is a connected graph with no cycles. To determine if the given graph is a valid tree, we can perform a depth-first search (DFS) or breadth-first search (BFS) traversal and check if there are any cycles and if all the nodes are reachable from a single source node.

Approach:
1. Build an adjacency list representation of the graph using the given edges.
2. Initialize a visited array to track the visited nodes during the graph traversal.
3. Perform a DFS or BFS traversal from any node in the graph:
   - Mark the current node as visited.
   - Visit all the neighbors of the current node:
     - If a neighbor is already visited and not the parent of the current node, it means there is a cycle, so return false.
     - If a neighbor is not visited, perform a recursive DFS or enqueue it in the BFS queue.
4. After the traversal, check if all the nodes are visited. If not, return false.
5. Return true if no cycle is found and all nodes are visited.

Time Complexity:
The time complexity depends on the graph traversal algorithm used. Using DFS or BFS, the time complexity is O(V + E), where V is the number of nodes (vertices) and E is the number of edges. We visit each node and edge once.

Space Complexity:
The space complexity is O(V + E), where V is the number of nodes (vertices) and E is the number of edges. This is the space used for the adjacency list and the visited array.
*/

class Solution {
  public:
    bool validTree(int n, vector<vector<int>> &edges) {
        vector<vector<int>> graph(n);   // Adjacency list representation of the graph
        vector<int> visited(n, 0);      // Visited array to track the visited nodes

        // Build the graph
        for (const auto &edge : edges) {
            int node1 = edge[0];
            int node2 = edge[1];
            graph[node1].push_back(node2);
            graph[node2].push_back(node1);
        }

        // Perform graph traversal to check for cycles and reachability
        if (!dfs(0, -1, graph, visited)) {
            return false;
        }

        // Check if all nodes are visited
        for (int i = 0; i < n; ++i) {
            if (visited[i] == 0) {
                return false;
            }
        }

        return true;
    }

  private:
    bool dfs(int node, int parent, vector<vector<int>> &graph, vector<int> &visited) {
        visited[node] = 1;  // Mark the current node as visited

        // Perform DFS traversal on the neighbors
        for (int neighbor : graph[node]) {
            if (visited[neighbor] == 0) {
                if (!dfs(neighbor, node, graph, visited)) {
                    return false;
                }
            } else if (neighbor != parent) {
                return false;
            }
        }

        return true;
    }
};

File: 11_Graphs/12_Graph_Valid_Tree/0261-graph-valid-tree.rs
/*
Problem: LeetCode 261 - Graph Valid Tree

Key Idea:
The key idea is to check whether the given undirected graph forms a valid tree. A valid tree has two properties:
1. It has n - 1 edges, where n is the number of nodes.
2. It is a connected graph, meaning all nodes are reachable from any node.

Approach:
1. Build an adjacency list representation of the undirected graph.
2. Initialize a set to keep track of visited nodes.
3. Perform a Depth-First Search (DFS) starting from node 0.
4. During DFS, check if we encounter any visited node that is not the parent of the current node (to avoid cycles).
5. After DFS, check if all nodes are visited. If not, return false (not connected).
6. Finally, check if the number of edges is n - 1. If true, return true; otherwise, return false.

Time Complexity:
O(V + E), where V is the number of nodes (vertices) and E is the number of edges in the graph.

Space Complexity:
O(V + E) for the adjacency list and visited set.
*/

impl Solution {
    pub fn valid_tree(n: i32, edges: Vec<Vec<i32>>) -> bool {
        let n = n as usize;
        let mut graph: Vec<Vec<usize>> = vec![vec![]; n];

        // Build the adjacency list
        for edge in edges.iter() {
            let u = edge[0] as usize;
            let v = edge[1] as usize;
            graph[u].push(v);
            graph[v].push(u);
        }

        let mut visited: Vec<bool> = vec![false; n];

        // Perform DFS to check connectivity
        if !Self::dfs(0, &mut visited, &graph, n, -1) {
            return false;
        }

        // Check if all nodes are visited
        if visited.iter().all(|&v| v) {
            return edges.len() == n - 1;
        }

        false
    }

    fn dfs(
        node: usize,
        visited: &mut Vec<bool>,
        graph: &Vec<Vec<usize>>,
        n: usize,
        parent: i32,
    ) -> bool {
        visited[node] = true;

        for &neighbor in graph[node].iter() {
            if !visited[neighbor] {
                if !Self::dfs(neighbor, visited, graph, n, node as i32) {
                    return false;
                }
            } else if neighbor as i32 != parent {
                return false;
            }
        }

        true
    }
}
